### PR TITLE
Rt1170 Dual Core and OpenAMP support

### PR DESCRIPTION
**Pull request by danieldegrasse**
Enable dual core and shared memory support for the RT1170 EVK.

This PR enables the following features for the RT1170:
- Dual core support (primary core can kick off secondary core without debugger)
- Shared memory support using `OCRAM2`
- Messaging unit driver to enable communication between cores

~Mulitcore images are created by converting the secondary image binary into a elf file with one section, and linking that into the primary image. The primary core will load the secondary image from flash into RAM, then boot the secondary core from that RAM.~
Mulicore images are now supported using Sysbuild

### CODE REVIEW

**Comment by carlocaione** [comment]
Just for the record: a lot of effort was lately done to move from IPM to MBOX, and from OpenAMP samples to IPC service.

**Comment by danieldegrasse** [comment]
@carlocaione Thanks for the heads up- I was unaware. Should I target enabling the `static_vrings_mi` sample on this board when I create an MBOX driver for it?

**Comment by carlocaione** [comment]
> @carlocaione Thanks for the heads up- I was unaware. Should I target enabling the `static_vrings_mi` sample on this board when I create an MBOX driver for it?

If you are willing to write a new MBOX drive I suggest you to enable the `samples/subsys/ipc/ipc_service/` sample. The `static_vrings_mi` is going to be deleted (in my TODO list :cry:  )

**Comment by dleach02** [comment]
@carlocaione is the project going to remove the IPM drivers? To be honest, I'm not up on the differences between IPM and MBOX but it seems like if the project has IPM support and MBOX support, it can be left up to the application to decide which way to enable this?

**Comment by carlocaione** [comment]
> @carlocaione is the project going to remove the IPM drivers?

No, this was discussed in the past and it was decided to keep both.

> To be honest, I'm not up on the differences between IPM and MBOX

MBOX supports multiple channels (for example several devices using the MBOX instance at the same time), IPM does not.

**Comment by danieldegrasse** [comment]
@carlocaione So to clarify, mbox drivers may use the same peripherals as ipm drivers, but they expose channels to allow multiple users of the IPM peripheral? In other words, is the mbox api essentially a new api to interact with the same peripheral as an ipm driver would use?

**Comment by carlocaione** [comment]
> So to clarify, mbox drivers may use the same peripherals as ipm drivers, but they expose channels to allow multiple users of the IPM peripheral? 

Correct.

> In other words, is the mbox api essentially a new api to interact with the same peripheral as an ipm driver would use?

Yes. The MBOX API was introduced because sometimes you want to share the peripheral between multiple users.  One notable example is when you have multiple OpenAMP instances that want to use the same mailbox peripheral for signalling. In that case you cannot use IPM but you have to use MBOX. For reference see:

https://github.com/zephyrproject-rtos/zephyr/blob/c9a3dae6519fc788d33233fc1c4d1364ba129034/samples/subsys/ipc/ipc_service/static_vrings/boards/nrf5340dk_nrf5340_cpuapp.overlay#L24-L56

In this case both (`ipc0` and `ipc1`) OpenAMP instances are using the same mbox peripheral (but different channels) at the same time.

File: soc/arm/nxp_imx/rt/mpu_regions.c
Comment (SebastianBoe) [inline soc/arm/nxp_imx/rt/mpu_regions.c:8]: I see you are not the first to need this header file.

Shall we do zephyr_include_directories(.) from soc/arm/common/cortex_m/CMakeLists.txt ?

Or perhaps even better move the header to

zephyr/include/arch/arm/aarch32/mpu/

```
sebo@sebo:~/ncs/bootloader/mcuboot$ ack arm_mpu_mem_cfg.h ~/ncs/
/home/sebo/ncs/zephyr/soc/arm/common/cortex_m/arm_mpu_regions.c
10:#include "arm_mpu_mem_cfg.h"

/home/sebo/ncs/zephyr/soc/arm/nuvoton_npcx/npcx7/mpu_regions.c
7:#include "../../common/cortex_m/arm_mpu_mem_cfg.h"

/home/sebo/ncs/zephyr/soc/arm/st_stm32/stm32h7/mpu_regions.c
8:#include "../../common/cortex_m/arm_mpu_mem_cfg.h"

```

Comment (carlocaione) [reply soc/arm/nxp_imx/rt/mpu_regions.c:8]: Also shameless plug for #41223 about this

File: soc/arm/nxp_imx/rt/CMakeLists.txt
Comment (tejlmand) [inline soc/arm/nxp_imx/rt/CMakeLists.txt:52]: This looks like an artificial dependency.

`app` is not really depending upon `remote_core_image` as far as I can tell.

**Review by tejlmand — CHANGES_REQUESTED** [review]
This looks like yet another multiimage approach.

Tagged for devreview.

Please take a look here #40555 and provide feedback so that we can solve the recurring multi image problem in a generic way suitable for all use-cases and not invent individual solutions everywhere.

File: samples/subsys/ipc/openamp/CMakeLists.txt
Comment (SebastianBoe) [inline samples/subsys/ipc/openamp/CMakeLists.txt:20]: @tejlmand : AFAICT this PR just adds support for another platform to do multi-image in a sample that already does multi-image.

So it is following prior-art and should not given -1 for doing yet-another-multi-image approach.

Comment (SebastianBoe) [reply samples/subsys/ipc/openamp/CMakeLists.txt:20]: reading this CMakeLists.txt file in it's entirety makes this clearer.

File: soc/arm/nxp_imx/rt/CMakeLists.txt
Comment (SebastianBoe) [reply soc/arm/nxp_imx/rt/CMakeLists.txt:51]: See https://github.com/zephyrproject-rtos/zephyr/pull/41579/files#r779630103

Comment (danieldegrasse) [reply soc/arm/nxp_imx/rt/CMakeLists.txt:51]: @tejlmand there is no current way to ensure `REMOTE_ZEPHYR_DIR` is set- but I think a build failure is the right way to approach the issue. I can add a check in the CMake file to provide a meaningful error to the user, but in the end selecting `CONFIG_DUAL_CORE_MCUX` tells the build system that there will be a second image present for the secondary core, so `REMOTE_ZEPHYR_DIR` must be set for the build to proceed.

File: soc/arm/nxp_imx/rt/CMakeLists.txt
Comment (danieldegrasse) [reply soc/arm/nxp_imx/rt/CMakeLists.txt:52]: It may be the wrong thing to depend on. The goal here is to force the primary image to depend on the remote image. Is there another target I could depend on besides `app` to ensure that dependency is enforced?

The reason for that requirement is that the final output dual core image is an image intended to run on the primary core, with the second core's image linked in as a section. This image is read from flash into RAM by the primary core, which then boots the secondary core from RAM.

File: samples/subsys/ipc/openamp/CMakeLists.txt
Comment (danieldegrasse) [reply samples/subsys/ipc/openamp/CMakeLists.txt:20]: @SebastianBoe that's correct, this PR is another multi-image approach. I do agree that we need a generic multi image approach in Zephyr, however.

Comment (tejlmand) [reply samples/subsys/ipc/openamp/CMakeLists.txt:20]: > So it is following prior-art and should not given -1 for doing yet-another-multi-image approach.

Except for the following two things I spotted immediately.
There might be more if taking a closer look:
- The SoC code itself relies on the `REMOTE_ZEPHYR_DIR` when `DUAL_CORE_MCUX` is enabled and then does a conversion and embed the second image into the Zephyr image.
  No other SoC code does this directly, as far as I can see.
  This is a new approach
- The PR introduces `DUAL_CORE_MCUX` Kconfig setting to control when a second image should be included.
   A very generic name, yet located specific in a Kconfig.soc:
   https://github.com/zephyrproject-rtos/zephyr/blob/c10d9c4d1e39c7e533e7f48c484d7a731c26d2ee/soc/arm/nxp_imx/rt/Kconfig.soc#L656
  This is a new strategy that I would like to be discussed, especially as we have #40555 coming.
    So a new setting that might be soon deprecated is not nice.

File: soc/arm/nxp_imx/rt/mpu_regions.c
Comment (danieldegrasse) [reply soc/arm/nxp_imx/rt/mpu_regions.c:8]: @SebastianBoe I agree that we should take one of those approaches with the file, but should it be something addressed by this PR, or a separate PR specifically calling out the issue and fixing it?

**Comment by danieldegrasse** [comment]
@tejlmand I've removed new KConfig symbols specific to this dual core implementation, and the soc level code now only builds for dual core if `CONFIG_OPENAMP` is selected.

File: soc/arm/nxp_imx/rt/mpu_regions.c
Comment (SebastianBoe) [reply soc/arm/nxp_imx/rt/mpu_regions.c:8]: A separate PR would be best.

File: samples/subsys/ipc/openamp/CMakeLists.txt
Comment (SebastianBoe) [reply samples/subsys/ipc/openamp/CMakeLists.txt:20]: > embed the second image into the Zephyr image.

I could have sworn this is what the open-amp sample has always been doing. They don't do hex/bin merging.
But perhaps this has changed in the last couple of years.

Comment (SebastianBoe) [reply samples/subsys/ipc/openamp/CMakeLists.txt:20]: If this is not just following prior-art then the author should at least state why it is necessary to do something different.

**Comment by Laczen** [comment]
@danieldegrasse, maybe a stupid question. Why is the image for the second core added to the flash of the primary ? Is it not possible to dedicate a piece of flash for the second core image ?

**Comment by dleach02** [comment]
> @tejlmand I've removed new KConfig symbols specific to this dual core implementation, and the soc level code now only builds for dual core if `CONFIG_OPENAMP` is selected.

I'm not sure `CONFIG_OPENAMP` is the right keyword to trigger on. We can certainly have a single embedded core on a larger system that enables openamp to talk to the larger (maybe Linux based) system. It feels like we want or need a config that is specific to multi-image for multi-CPU systems and not necessarily tie it to whether or not OpenAMP is used.

**Comment by danieldegrasse** [comment]
@Laczen The requirement to run the second core from RAM means that the second image must be linked with the first. This is because the primary core image must know the size of the second core image in order to copy it from flash to ram.

To be clear, it is possible to boot the secondary core directly from flash, but this is not preferred because when both the M7 and M4 cores share access to the flash it will result in delayed instruction fetches and slower execution. Moreover, if both cores share access to the flash, there would need to be additional application code to handle flash writes, since one core writing to the flash while the other reads would create a RWW hazard

**Comment by tejlmand** [comment]
@danieldegrasse based on the talks at dev-review meeting I have created a small PR which allows to generate a header file containing the information you said was needed to know for the first core.

Please take a look here: #41685.

This generates a header with the information you should need in the form:
```
define SEGMENT_NUM 4

#define SEGMENT_LMA_ADDRESS_0 0x20205be0
#define SEGMENT_VMA_ADDRESS_0 0x20205be0
#define SEGMENT_SIZE_0 0x8

#define SEGMENT_LMA_ADDRESS_1 0x20200000
#define SEGMENT_VMA_ADDRESS_1 0x20200000
#define SEGMENT_SIZE_2 0x6f3c

.....
```
[UTIL_LISTIFY](https://docs.zephyrproject.org/latest/reference/util/index.html#c.UTIL_LISTIFY) is useful to create an array of this data.

Please provide feedback as to how this would fit your needs.
If it does, I think #41685 will be useful together with #40555.

**Comment by tejlmand** [comment]
> > It feels like we want or need a config that is specific to multi-image for multi-CPU systems and not necessarily tie it to whether or not OpenAMP is used.
>
> @tejlmand, thoughts?

Correct, we should get a lot of this knowledge out of the single image Zephyr build and then have a higher level doing the multi image work, which is what i'm working on here #40555.
This is part of higher level work here: https://github.com/zephyrproject-rtos/zephyr/issues/41640 which tries to track the various requirements needed.

Until that work is complete, we should try to simply contain current multi-image solutions inside `samples/subsys/ipc` and not have it spread to samples in general.
Only reason to still allow `subsys/ipc` to do custom multi-image is because they have been allowed in past where we had no general solution available.

File: samples/subsys/ipc/openamp/CMakeLists.txt
Comment (tejlmand) [reply samples/subsys/ipc/openamp/CMakeLists.txt:20]: > > embed the second image into the Zephyr image.
> 
> I could have sworn this is what the open-amp sample has always been doing. They don't do hex/bin merging. But perhaps this has changed in the last couple of years.

This is the mess I would like to get cleaned up by having a uniform multi-image approach.
Unified in the sense it is not tied into single image Zephyr build.
Different multi-core SoC of course may have different ways they need to flash / embed multiple images.

So currently for `openamp` we have:
- lpcxpresso54114_m4 and lpcxpresso55s69_cpu0 which uses `generate_inc_file_for_gen_target()` and has a dedicated `SECOND_CORE_MCUX` setting
- mps2_an521 and v2m_musca_b1 which don't embed the second image, and don't have special setting.
   mps2_an521 even with the twist that if you build for cpu1 and not `openamp` then you always get a dummy image build for cpu0 that you cannot change. Not setting to disable and no direct Kconfig options for that image.

and now this PR which embeds the second image in another way not covered by the above methods but takes the bin file directly and embed as a section.

I prefer at this point we get a higher level handling of multi-image and thus spend time on such solution instead of yet another small variant for another SoC / board under `subsys/ipc`.

**Comment by danieldegrasse** [comment]
@tejlmand I've reworked this PR based off your work in #41685. The SoC level code still utilizes `CONFIG_OPENAMP` to determine if the M4 core should be booted from the M7, but I moved cmake changes into the openamp sample where possible. Please let me know your thoughts on this approach.

**Comment by dleach02** [comment]
Daniel, the HAL PR was merged so you can update the west.yaml.

**Comment by danieldegrasse** [comment]
@dleach02 I've updated the PR to use the messaging unit driver introduced with #42320

**Comment by dleach02** [comment]
HAL PR was merged so this needs a change to the west.yaml

**Comment by danieldegrasse** [comment]
@dleach02 rebased to include latest `west.yml`

**Comment by dleach02** [comment]
@tejlmand, @carlocaione can you revisit

File: soc/arm/nxp_imx/rt/CMakeLists.txt
Comment (danieldegrasse) [reply soc/arm/nxp_imx/rt/CMakeLists.txt]: I updated the PR to use that. Note- the MPU region attribute for the shared memory needs to be `REGION_IO_ATTR` instead of `REGION_RAM_NOCACHE_ATTR`, so I updated the DT binding for `zephyr,memory-region` to allow for "IO" to be one of the possible `zephyr,memory-region-mpu` settings.

**Review by carlocaione — DISMISSED** [review]
I'm not an expert on this platform but my comments were addressed.

**Comment by danieldegrasse** [comment]
@tejlmand Can you revisit this PR? Given the fact that separate images are being built, and there is not any custom solution for `west flash` used by this PR, I think that transitioning this multi image support from its current state to something compatible with sysbuild will be relatively simple when sysbuild merges. In the meantime, I'd really like to see this PR go in: It will provide a sample for customers showing how to boot the M4 core present on the RT1170, which they can extend for their specific use case.

**Comment by mbolivar-nordic** [comment]
@danieldegrasse unfortunately @tejlmand is out on parental leave and won't be back in time for code freeze. If you want this for v3.1, I suggest you apply for an exception from the TSC to see if dismissing his review and merging for v3.1 after code freeze is acceptable.

**Comment by danieldegrasse** [comment]
@mbolivar-nordic I'll discuss with @dleach02 to see if that is something we'd like to do, thanks for the heads up

**Comment by tyalie** [comment]
Are there any news on this now that #40555 has been merged?

**Comment by danieldegrasse** [comment]
This PR is dependent on #48709, which itself depends on #49552. I've kept this PR open in the meantime because I would like to have it available as a demonstration of how to support multicore operation on the RT11xx series. However, it may be some time before this is able to merge.

**Review by tejlmand — APPROVED** [review]
lgtm

**Review by dleach02 — APPROVED** [review]

**Pull request merged**
